paywall/src/unlock.js/Wallet.ts

```diff
@@ -0,0 +1,315 @@
+import { Web3Window, web3MethodCall } from '../windowTypes'
+import { PostMessages } from '../messageTypes'
+import IframeHandler from './IframeHandler'
+import { PaywallConfig } from '../unlockTypes'
+import StartupConstants from './startupTypes'
+
+declare const process: {
+  env: any
+}
+
+/**
+ * This class handles everything relating to the web3 wallet, including key purchases
+ *
+ * It either sets up a proxy wallet to the user wallet,
+ * or a proxy wallet for user accounts.
+ * If user accounts will be used, it is explicitly
+ * specified in the configuration, and a crypto
+ * wallet is not present. In that situation,
+ * the wallet will trigger a load of the
+ * user accounts iframe and initialize it
+ */
+export default class Wallet {
+  private readonly iframes: IframeHandler
+  private readonly window: Web3Window
+  private readonly hasWallet: boolean = true
+  private readonly isMetamask: boolean
+  private readonly config: PaywallConfig
+  private hasWeb3: boolean = false
+  private useUserAccounts: boolean = false
+
+  private userAccountAddress: string | null = null
+  private userAccountNetwork: number
+
+  constructor(
+    window: Web3Window,
+    iframes: IframeHandler,
+    config: PaywallConfig,
+    constants: StartupConstants
+  ) {
+    this.window = window
+    this.iframes = iframes
+    this.config = config
+    this.userAccountNetwork = constants.network
+
+    // do we have a web3 wallet?
+    this.hasWallet = !!(this.window.web3 && this.window.web3.currentProvider)
+    this.isMetamask = !!(
```

Comment:
`this.hasWallet && window.web3.currentProvider.isMetamask` ?

Reply:
ah, I remember why I did it this way - typescript is not smart enough to recognize that `this.hasWallet` implies the existence of both `web3` and `web3.currentProvider`

I need to make a window type that has web3 defined to solve some of these little issues